File: kernel/internal/ledger/attest.go
package ledger

// C5 — the inference attestation.
//
// C4 answers "what acted on the world, and who authorized it". It does not
// answer the question that comes immediately after in any real incident
// review: **what produced the judgement that led to the act?** A sealed entry
// saying `motor.erp.invoice.create` was delivered under policy 7c1e… is
// evidence about the effect and says nothing about the model whose output
// argued for it — which model, which weights, which quantization, which
// sampling parameters, which seed. Those are exactly the variables that decide
// whether a bad outcome was a policy failure, a prompt failure, or a silently
// swapped model.
//
// An attestation is one skill's assertion about how it produced an output. The
// kernel content-addresses it, stores it once, and binds its hash into every
// effect that the attested output causally led to. What that buys:
//
//   - "This invoice was created on the strength of a reply from
//     Qwen2.5-1.5B-Instruct at Q4_K_M, revision f1d2…, temperature 0.7,
//     seed 42" becomes a checkable claim rather than a story.
//   - Swapping a model under a running graph changes the attestation hash,
//     which changes every subsequent ledger entry, which breaks the chain
//     unless it is sealed honestly. A silent model swap stops being silent.
//   - Replay (`aura replay`) can compare not just outputs but the inference
//     conditions that produced them, which is what makes a diff meaningful
//     when sampling is stochastic.
//
// ## What this proves, and what it does not
//
// Stated plainly, because the difference decides whether a reader is misled:
// **an attestation is an assertion by the skill, cryptographically bound to
// the effects it led to and to the moment it was made — it is not proof that
// the skill told the truth.** A skill that lies about which model it ran
// produces a sealed, chained, signed record of a lie. What the kernel
// guarantees is narrower and still useful: the assertion cannot be altered
// afterwards, cannot be detached from the effects it caused, and cannot be
// backdated. Making the assertion itself trustworthy needs hardware
// attestation (a TEE quote binding the measurement of the process that ran
// the model), which C5 leaves room for in `tee` and the roadmap tracks
// separately.
//
// The honest one-line summary, which the README and `aura receipt` both use:
// *who claimed what, when, and what it caused — verifiable; whether the claim
// was true — only as far as you trust the skill that made it.*

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"strings"
	"time"

	"aura/kernel/internal/spec"
	"aura/kernel/internal/store"
)

// MaxAttestationBytes bounds one attestation record. Attestations are
// metadata about an inference, never its content: a few hundred bytes of
// model identity and sampling parameters. The cap exists so a skill cannot
// use the field as an unbounded, permanently-retained side channel into the
// node's storage — the same reasoning that keeps payloads out of the ledger.
const MaxAttestationBytes = 8 << 10

// Attestation is the parsed view of what a skill asserted. The raw bytes are
// what gets hashed and stored; this struct exists so the kernel can validate
// the shape, and so `aura bom` and `aura receipt` can read it without
// re-deriving the schema.
type Attestation struct {
	// Engine identifies the runtime that executed the inference —
	// "llama.cpp", "faster-whisper", "piper", "openai". Mandatory: an
	// attestation that does not say what ran it is not evidence of anything.
	Engine        string `json:"engine"`
	EngineVersion string `json:"engine_version,omitempty"`

	// Model is the model's identity in whatever namespace the engine uses.
	// For anything from the Hugging Face Hub this is the repo id
	// ("Qwen/Qwen2.5-1.5B-Instruct-GGUF"), which is what makes ModelRevision
	// meaningful and what lets `aura bom` emit a resolvable component.
	Model string `json:"model"`
	// ModelRevision is the immutable revision the weights came from — an HF
	// commit sha. A repo id alone names a moving target; the pair names one
	// specific set of bytes, which is the only version of the claim worth
	// sealing.
	ModelRevision string `json:"model_revision,omitempty"`
	// ModelFile and ModelSHA256 pin the artifact itself, for the local case
	// where the weights are a file on disk rather than a Hub reference.
	ModelFile   string `json:"model_file,omitempty"`
	ModelSHA256 string `json:"model_sha256,omitempty"`
	// Quantization ("Q4_K_M", "fp16", "int8") is separated from the file name
	// because it changes the output distribution and is the single most
	// common undeclared difference between "the same model" in two places.
	Quantization string `json:"quantization,omitempty"`

	// Params carries the sampling configuration — temperature, top_p, top_k,
	// max_tokens, and crucially `seed`. Left as raw JSON so the kernel does
	// not have to model every engine's knobs, and so the hash covers exactly
	// what the skill sent.
	Params json.RawMessage `json:"params,omitempty"`

	// PromptSHA256 binds the attestation to the input it was made about,
	// without storing the prompt: the same reason the ledger stores
	// payload_sha256 and never the payload. A prompt is user data and must
	// not become permanently undeletable.
	PromptSHA256 string `json:"prompt_sha256,omitempty"`
	// OutputSHA256 does the same for what came back.
	OutputSHA256 string `json:"output_sha256,omitempty"`

	// TEE is reserved for a hardware attestation quote (Intel TDX, AMD
	// SEV-SNP, NVIDIA CC) binding this record to a measured enclave. Absent
	// today on every node; declared now so its later arrival is additive
	// rather than a new major. Its absence is what makes everything above an
	// assertion rather than a proof — see the package comment.
	TEE json.RawMessage `json:"tee,omitempty"`

	// Energy is what the skill measured or estimated for this inference.
	// Optional, and honest about which of the two it is — see EnergyReport.
	Energy *EnergyReport `json:"energy,omitempty"`
}

// EnergyReport carries the cost of an inference in millijoules, together with
// how the number was arrived at.
//
// `Source` is mandatory and not decorative. A measured joule and a modelled
// joule differ by an order of magnitude in trustworthiness, and a field that
// reported only the number would quietly launder an estimate into a
// measurement the first time anyone put it in a sustainability report. A
// consumer that only accepts hardware counters can filter on this; one that
// accepts estimates knows what it is accepting.
type EnergyReport struct {
	Millijoules float64 `json:"millijoules"`
	// Source is one of the EnergySources values: what produced this number.
	Source string `json:"source"`
	// Basis explains an estimate's inputs in one line ("14.2s wall x 45W
	// declared TDP"), so a reader can judge it instead of trusting it. Empty
	// for a hardware measurement, where the counter is the basis.
	Basis string `json:"basis,omitempty"`
}

// EnergySources are the ways an energy figure can come to exist, ordered from
// most to least trustworthy. Generated from spec/enums.yaml, like every other
// contract enumeration — a list duplicated here would be a list that drifts
// from the Python SDK's copy without anything failing.
var EnergySources = spec.EnergySources

// Hash content-addresses an attestation: sha256 over exactly the bytes the
// skill sent.
//
// Over the raw bytes, not over a re-marshalling of the parsed struct, for the
// same reason LeafHash hashes stored bytes: any canonicalization step is a
// place where the sealer's encoder and a verifier's encoder can differ, and
// Params/TEE are free-form JSON where that risk is real rather than
// theoretical.
func AttestationHash(raw []byte) string {
	sum := sha256.Sum256(raw)
	return "sha256:" + hex.EncodeToString(sum[:])
}

// ParseAttestation validates a skill's assertion and returns both the parsed
// view and the canonical hash of the raw bytes.
//
// Validation is deliberately thin — engine and model must be present and the
// record must fit — because the kernel is not in a position to judge whether
// a claim is *true*, only whether it is well-formed enough to be evidence.
// Rejecting a malformed one matters anyway: an attestation that cannot be
// parsed later is a hash bound into the chain pointing at nothing readable.
func ParseAttestation(raw []byte) (Attestation, string, error) {
	var a Attestation
	if len(raw) == 0 {
		return a, "", fmt.Errorf("attestation is empty")
	}
	if len(raw) > MaxAttestationBytes {
		return a, "", fmt.Errorf(
			"attestation is %d bytes, over the %d-byte cap — attestations carry model "+
				"identity and sampling parameters, never prompts or outputs",
			len(raw), MaxAttestationBytes)
	}
	if err := json.Unmarshal(raw, &a); err != nil {
		return a, "", fmt.Errorf("attestation is not valid JSON: %w", err)
	}
	if strings.TrimSpace(a.Engine) == "" {
		return a, "", fmt.Errorf("attestation declares no engine")
	}
	if strings.TrimSpace(a.Model) == "" {
		return a, "", fmt.Errorf("attestation declares no model")
	}
	if a.Energy != nil {
		if a.Energy.Source == "" {
			return a, "", fmt.Errorf(
				"attestation reports energy with no source — a number whose provenance is " +
					"unstated cannot be told apart from a guess")
		}
		if !oneOf(a.Energy.Source, EnergySources) {
			return a, "", fmt.Errorf("attestation energy source %q is not one of %v",
				a.Energy.Source, EnergySources)
		}
	}
	return a, AttestationHash(raw), nil
}

// RecordAttestation stores one attestation, content-addressed. Storing the
// same record twice is a no-op: identical inference conditions across a
// thousand deliveries are one row, which is what keeps a busy node's storage
// proportional to its *distinct* configurations rather than to its traffic.
func RecordAttestation(st *store.Store, hash string, raw []byte) error {
	return st.SaveAttestation(hash, raw, time.Now().UnixMilli())
}

// LoadAttestation reads one back by hash and re-verifies the content address
// before returning it.
//
// Re-hashing on read is not paranoia: the attestation table is outside the
// Merkle tree (it holds records, not ledger entries), so its integrity rests
// entirely on the hashes the sealed entries cite. Checking here is what makes
// that binding load-bearing rather than decorative — an edited attestation
// record fails to load rather than quietly returning different claims than
// the ones the chain committed to.
func LoadAttestation(st *store.Store, hash string) (Attestation, error) {
	raw, err := st.Attestation(hash)
	if err != nil {
		return Attestation{}, err
	}
	if got := AttestationHash(raw); got != hash {
		return Attestation{}, fmt.Errorf(
			"attestation %s does not match its content address (recomputes to %s) — "+
				"the stored record was altered after the ledger cited it", hash, got)
	}
	a, _, err := ParseAttestation(raw)
	return a, err
}

File: sdk/python/src/aura/attest.py
"""C5 — inference attestation: what a skill asserts about how it produced an output.

A skill that runs a model can say so, in a form the kernel binds to every
effect that output leads to:

    from aura import Attestation, Context, Skill

    @skill.on("text_in")
    async def handle(ctx: Context) -> None:
        reply = run_the_model(ctx.payload["text"])
        await ctx.emit(
            "text_out",
            {"text": reply, "final": True},
            attest=Attestation(
                engine="llama.cpp",
                model="Qwen/Qwen2.5-1.5B-Instruct-GGUF",
                model_revision="f1d2d2f924e986ac86fdf7b36c94bcdf32beec15",
                quantization="Q4_K_M",
                params={"temperature": 0.7, "seed": 42},
            ),
        )

The kernel hashes the record, stores it once, and cites the hash in the ledger
entry of any `motor.*` effect downstream. An auditor reading that entry later
can therefore ask not only "who authorized this" but "what argued for it".

## What this proves

The honest version, because the overclaim is tempting and would be caught:
**an attestation is an assertion, cryptographically bound to what it caused
and to when it was made — not proof that the assertion is true.** A skill that
lies about its model produces an unforgeable record of a lie. What the kernel
guarantees is that the claim cannot be altered afterwards, cannot be detached
from the effects it led to, and cannot be backdated. Making the claim itself
trustworthy needs hardware attestation, which the `tee` field is reserved for.

Attesting is optional and additive. A skill that never calls it behaves
exactly as it did before C5 existed.
"""
from __future__ import annotations

import hashlib
import json
import logging
import os
import shutil
import subprocess
import time
from dataclasses import dataclass, field
from typing import Any

from ._spec import ENERGY_SOURCES  # generated from spec/enums.yaml

_logger = logging.getLogger("aura.attest")

# Mirrors ledger.MaxAttestationBytes. Attestations carry model identity and
# sampling parameters — never prompts, never outputs.
MAX_ATTESTATION_BYTES = 8 * 1024


@dataclass
class Energy:
    """The energy one inference cost, and how that number came to exist.

    `source` is mandatory and load-bearing. A measured joule and a modelled
    joule differ by an order of magnitude in trustworthiness, and a field
    carrying only the number would launder an estimate into a measurement the
    first time it reached a sustainability report. Consumers that only accept
    hardware counters can filter; consumers that accept estimates know what
    they are accepting.
    """

    millijoules: float
    source: str
    basis: str = ""

    def to_wire(self) -> dict:
        if self.source not in ENERGY_SOURCES:
            raise ValueError(
                f"energy source {self.source!r} must be one of {ENERGY_SOURCES}")
        out: dict[str, Any] = {
            "millijoules": round(float(self.millijoules), 3),
            "source": self.source,
        }
        if self.basis:
            out["basis"] = self.basis
        return out


@dataclass
class Attestation:
    """What a skill claims about the inference behind an output."""

    engine: str
    model: str
    engine_version: str = ""
    model_revision: str = ""
    model_file: str = ""
    model_sha256: str = ""
    quantization: str = ""
    params: dict[str, Any] = field(default_factory=dict)
    prompt_sha256: str = ""
    output_sha256: str = ""
    energy: Energy | None = None
    tee: Any = None

    def to_wire(self) -> dict:
        if not self.engine:
            raise ValueError("an attestation must name the engine that ran the inference")
        if not self.model:
            raise ValueError("an attestation must name the model")
        wire: dict[str, Any] = {"engine": self.engine, "model": self.model}
        for key in (
            "engine_version", "model_revision", "model_file", "model_sha256",
            "quantization", "prompt_sha256", "output_sha256",
        ):
            value = getattr(self, key)
            if value:
                wire[key] = value
        if self.params:
            wire["params"] = self.params
        if self.energy is not None:
            wire["energy"] = self.energy.to_wire()
        if self.tee is not None:
            wire["tee"] = self.tee

        encoded = json.dumps(wire, separators=(",", ":"), sort_keys=True)
        if len(encoded.encode("utf-8")) > MAX_ATTESTATION_BYTES:
            raise ValueError(
                f"attestation is over the {MAX_ATTESTATION_BYTES}-byte cap — it carries "
                "model identity and sampling parameters, never prompts or outputs")
        return wire


def sha256_text(text: str) -> str:
    """Hash a prompt or an output for `prompt_sha256` / `output_sha256`.

    Binds an attestation to the input it was made about without storing that
    input — the same reason the ledger keeps `payload_sha256` and never the
    payload. A prompt is user data and must not become permanently
    undeletable.
    """
    return hashlib.sha256(text.encode("utf-8")).hexdigest()


def sha256_file(path: str, *, max_bytes: int = 0) -> str:
    """Hash a weights file, so `model_sha256` pins the actual artifact.

    `max_bytes` caps how much is read. Hashing a 40 GB GGUF on every startup
    is not free, and a caller that would rather pin the first N bytes than pin
    nothing can say so — the field is a fingerprint, and a partial one that
    exists beats a complete one nobody computes. When capped, the caller
    should say so in the model_file name or skip the field; a truncated hash
    presented as a whole-file hash would be a quiet lie.
    """
    h = hashlib.sha256()
    read = 0
    with open(path, "rb") as fh:
        while True:
            chunk = fh.read(1 << 20)
            if not chunk:
                break
            if max_bytes and read + len(chunk) > max_bytes:
                h.update(chunk[: max_bytes - read])
                break
            h.update(chunk)
            read += len(chunk)
    return h.hexdigest()


class EnergyMeter:
    """Measures the energy an inference cost, or estimates it and says so.

    Tries hardware counters first and falls back to an analytical estimate,
    never silently: `Energy.source` records which happened, and an estimate
    carries the inputs it was computed from so a reader can judge it rather
    than trust it.

    Usage::

        with EnergyMeter() as meter:
            reply = run_the_model(prompt)
        attestation.energy = meter.result()

    On a machine with no readable counters this still produces a number, and
    that number is explicitly labelled `estimated`. The alternative — omitting
    energy on most hardware — would mean the field only ever appears on
    datacentre GPUs, which is where it matters least.
    """

    # Fallback power draw when nothing can be measured, in watts. Deliberately
    # a whole-package figure for a busy consumer machine rather than a precise
    # one: an estimate's job here is an order of magnitude plus an honest
    # label, and a suspiciously precise default would invite the false
    # confidence the `source` field exists to prevent.
    DEFAULT_WATTS = 45.0

    def __init__(self, *, fallback_watts: float | None = None) -> None:
        self._watts = fallback_watts or self.DEFAULT_WATTS
        self._start = 0.0
        self._elapsed = 0.0
        self._start_energy: tuple[str, float] | None = None
        self._end_energy: tuple[str, float] | None = None

    def __enter__(self) -> "EnergyMeter":
        self._start = time.perf_counter()
        self._start_energy = _read_energy_counter()
        return self

    def __exit__(self, *exc: object) -> None:
        self._elapsed = time.perf_counter() - self._start
        self._end_energy = _read_energy_counter()

    def result(self) -> Energy:
        start, end = self._start_energy, self._end_energy
        if start and end and start[0] == end[0] and end[1] >= start[1]:
            source, joules = end[0], end[1] - start[1]
            # A counter that did not move over a real interval is a counter
            # that is not actually reporting; fall through to the estimate
            # rather than claim a measured zero.
            if joules > 0:
                return Energy(millijoules=joules * 1000.0, source=source)

        joules = self._watts * self._elapsed
        return Energy(
            millijoules=joules * 1000.0,
            source="estimated",
            basis=f"{self._elapsed:.2f}s wall x {self._watts:.0f}W assumed package draw",
        )


def _read_energy_counter() -> tuple[str, float] | None:
    """Read a cumulative energy counter in joules, if the platform has one."""
    for reader in (_read_rapl, _read_nvml):
        try:
            value = reader()
        except Exception:  # noqa: BLE001 - a counter that errors is a counter we do not have
            continue
        if value is not None:
            return value
    return None


def _read_rapl() -> tuple[str, float] | None:
    """Intel/AMD RAPL, exposed by Linux as a microjoule counter in sysfs."""
    path = "/sys/class/powercap/intel-rapl:0/energy_uj"
    if not os.path.exists(path):
        return None
    with open(path) as fh:
        return ("rapl", int(fh.read().strip()) / 1_000_000.0)


def _read_nvml() -> tuple[str, float] | None:
    """NVIDIA total energy consumption, via nvidia-smi.

    Queried through the CLI rather than the NVML bindings on purpose: the SDK
    has no runtime dependencies and is not going to acquire one for an
    optional field. `nvidia-smi` is present wherever the driver is.
    """
    exe = shutil.which("nvidia-smi")
    if not exe:
        return None
    out = subprocess.run(  # noqa: S603 - fixed executable, fixed arguments
        [exe, "--query-gpu=total_energy_consumption", "--format=csv,noheader,nounits"],
        capture_output=True, text=True, timeout=5, check=False,
    )
    if out.returncode != 0:
        return None
    first = out.stdout.strip().splitlines()
    if not first:
        return None
    # Reported in millijoules.
    return ("nvml", float(first[0].strip()) / 1000.0)
